keps/sig-cluster-lifecycle/wgs/20191001-component-config-validation-interface.md

```diff
@@ -0,0 +1,177 @@
+- A required system setting or module may not be setup properly or available, hence, prohibiting component operation with the current config.
+
+Unfortunately, due to the nature of Kubernetes, such kinds of failures might not be detected by cluster lifecycle tools early enough and users might not be warned in time, that their configs are invalid.  
+Currently, some cluster lifecycle tools vendor in and use Go code from the target components themselves to be able to verify component configs early enough during the provisioning operation, so that meaningful error messages are displayed in time directly to the user. This, however, poses its own problems as such pieces of code can deviate from the actual shipped component and can lead to false positives and false negatives.  
+Hence, it’s best that validation is performed by the component itself in a one off operation.
+
+
+### Goals
+
+The main goal is to provide a stable standard command line interface for components that can be used to validate component configs and retrieve appropriate errors and warnings.
+
+
+### Non-Goals
+
+To impose the implementation of this proposal on all Kubernetes components, that provide component config. Implementing this interface is entirely optional.
+
+
+## Proposal
+
+The proposal is to introduce a new standardized sub-command for each component that implements the interface. This command can then be called by invoking the component’s executable or creating a container with it.
+
+
+### The “experimental-config-validate“ sub-command
+
+The new sub-command will be known as “experimental-config-validate”. Upon graduation to beta and GA, this sub-command will be renamed to just “config-validate”.
+
+It must support a couple of mandatory flags:
+- “--help” or “-h” - if specified, will dump usage information of the sub-command and exit without doing any config validation.
+- “--file” or “-f” - supplies a YAML or JSON file containing a component config to be validated.
+
+In all implementations of “experimental-config-validate”, the “--file” flag can be specified multiple times. If a component does not support multi-file config, supplying multiple “--file” values must return an appropriate error to the user without performing any validation.  
+If no “--file” or “--help” switches are supplied, the command reads the config to be validated from the standard input.
+
+When the “experimental-config-validate” is executed supplied with a config, it should validate it and:
+- Exit with success if the supplied config is found to be valid. Each line of text printed on the standard output is considered to be a single warning message. Multiple warning messages can be printed.
+- Exit with failure if an error is encountered and the component will be unable to operate normally with the config as is. Each line of text printed on the standard output is considered to be a single error message. Multiple error messages can be printed.
+
```

Comment:
the proposal of a new command knowing about all components sounds like yet another hyperkube. We are on the way to move hyperkube out of k/k to cut dependencies. With this we will get that back? /cc @dims